Document new linker options and trimming libraries

sbomer opened the pull request:
This adds documents for the new linker options and instructions for trimming libraries. The "Trimming libraries" document will be the target of the aka.ms link in [IL2104](https://github.com/mono/linker/blob/main/docs/error-codes.md#il2104-assembly-assembly-produced-trim-warnings-for-more-information-see-httpsakamsdotnet-illinklibraries) and should (hopefully) nicely complement the [blog post](https://github.com/mono/linker/pull/1946) @agocke. For now this mostly uses similar examples to the blog post, but over time I think this would be a good place to add more involved examples that could be useful to developers.

For trimming libraries I think we should fix https://github.com/dotnet/sdk/issues/16140 to get the behavior described in the doc.

This also fixes https://github.com/dotnet/docs/issues/23740.

### Internal preview

✔️ [Prepare libraries for trimming](https://review.docs.microsoft.com/en-us/dotnet/core/deploying/prepare-libraries-for-trimming?branch=pr-en-us-23766)
✔️ [Trimming options](https://review.docs.microsoft.com/en-us/dotnet/core/deploying/trimming-options?branch=pr-en-us-23766)

sbomer committed:
Document new linker options and trimming libraries

sbomer committed:
Fix TOC

sbomer committed:
Document EnableTrimAnalyzer

And fix typos, links

sbomer committed:
PR feedback

- Move sample up
- Call out TrimMode link default
- Clarify app vs library
- Publish Release

sbomer commented on docs/core/deploying/trimming-libraries.md:
```diff
@@ -0,0 +1,161 @@
+
+## Trim warnings for libraries
+
+`PublishTrimmed` will produce trim analysis [warnings](trimming-options.md#analysis-warnings) for patterns that are not statically understood to be compatible with trimming, including patterns in your code and in dependencies. You may see warnings like `warning IL2104: Assembly 'SomeAssembly' produced trim warnings`.
```
Yeah, it was supposed to give some context on where warnings like IL2104 come from. I reorganized it a bit to make it more clear, PTAL

sbomer committed:
Change title to "Preparing libraries for trimming"

tlakollo commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,163 @@
+
+- `<TrimmerDefaultAction>link</TrimmerDefaultAction>` ensures that only used parts of dependencies are analyzed. Without this option, you would see warnings originating from _any_ part of a dependency that doesn't set `[AssemblyMetadata("IsTrimmable", "True")]`, including parts that are unused by your library.
+
+- `<SuppressTrimAnalysisWarnings>link</SuppressTrimAnalysisWarnings>` is implied by `PublishTrimmed` in the .NET SDK. It is only needed if the app targets form factors like Android where the trim analysis warnings are disabled by default. Consult the documentation for your SDK.
```
Nit: this should be <SuppressTrimAnalysisWarnings>false<SuppressTrimAnalysisWarnings>

sbomer replied:
Good catch, thanks!

tlakollo commented on docs/core/deploying/preparing-libraries-for-trimming.md:
````diff
@@ -0,0 +1,163 @@
+}
+```
+
+Here, `Foo` is calling a method which takes a `Type` argument that is annotated with a [`DynamicallyAccessedMembers`](https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute?view=net-5.0) requirement. The requirement states that the value passed in as this argument must represent a type whose public methods are available. In this case, you can fix ...
````
Maybe we should put another piece of code with the code changes for this and the other examples a before/after to be more explicit.
Also, I'm not comfortable with using "Like above, once you have bubbled ..." If we update the document we might point to something else, better to put something like "Like with RequiresUnreferencedCode, once you have bubbled ..."

tlakollo commented on docs/core/deploying/preparing-libraries-for-trimming.md:
````diff
@@ -0,0 +1,163 @@
+```
+
+This means the library calls a method which has explicitly been annotated as incompatible with trimming, using [`RequiresUnreferencedCodeAttribute`](
+https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.requiresunreferencedcodeattribute?view=net-5.0). To get rid of the warning, consider whether `Foo` needs to call `Bar` to do its job. If so, annotate the caller `Foo` with `RequiresUnreferencedCode` as well; this will "bubble up" the warning so that callers of `Foo` get a warning instead. Once you have "bubbled up" the attribut ...
````
Is this the right place to include the DynamicDependency, Descriptor options to solve the RequiresUnreferencedCode warning and be able to use UnconditionalSuppressMessage?

sbomer replied:
Yes, I think either in this doc or one next to it would be a good place for these. I might do it in a separate PR if I don't get to it soon.

eerhardt replied:
When we add this, we should have a "warning" or "advanced" banner on it. It is definitely a "this is something you can do to make things work, but it is really only for people who know precisely what they are doing".

sbomer replied:
Added a new section with UnconditionalSuppressMessage. I'm having some trouble coming up with examples for DynamicDependency that aren't super contrived. Do you have any suggestions?

eerhardt replied:
> Do you have any suggestions?

One example is when serialization doesn't work. 
* https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Runtime.Serialization.Xml/tests/TrimmingTests/EndToEndTest.cs#L13-L15
* https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L177-L182

sbomer replied:
Thanks! I will add an example for DynamicDependency in a follow-up.

sbomer committed:
PR feedback

- Fix typo
- Avoid referencing sections with "above/below"

sbomer committed:
PR feedback

Add code snippet for examples after annotation

eerhardt commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,188 @@
+
+You will encounter detailed warnings originating from your own code and `ProjectReference` dependencies. You may also see warnings like `warning IL2104: Assembly 'SomeAssembly' produced trim warnings` for `PackageReference` libraries. This means that the library contained patterns which are not guaranteed to work in the context of the trimmed app, and may result in a broken app. Consider contacti ...
+
+To resolve warnings originating from the app code, jump ahead to the instructions on [resolving trim warnings](resolving-trim-warnings.md). If you are interested in making your own `ProjectReference` libraries trim friendly, follow the instructions to [enable library trim warnings](enabling-library-trim-warnings).
```
Both links in this line 404 for me.

eerhardt commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,188 @@
+
+These instructions show how to enable and resolve static analysis warnings to prepare a library for trimming. Follow these steps if you are authoring a library and either want to proactively make your library trimmable, or have been contacted by app authors who encountered trim warnings from your library.
+
+Use the .NET 6 SDK for the best experience. It is possible to [show analysis warnings](trimming-options.md#analysis-warnings) in .NET 5, but this will show detailed warnings for every library, including framework libraries. These instructions assume you are using the .NET 6 SDK.
```
```suggestion
Use the .NET 6 SDK for the best experience. It is possible to [show analysis warnings](trimming-options.md#analysis-warnings) in .NET 5, but not all framework assemblies were annotated correctly in .NET 5. Also, this will show detailed warnings for every library, including framework libraries. These instructions assume you are using the .NET 6 SDK.
```

eerhardt commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,188 @@
+
+Note that the analysis results depend on the implementation details of your dependencies. If you update to a new version of a dependency, this may introduce analysis warnings if the new version added non-understood reflection patterns, even if there were no API changes. In other words, introducing trim analysis warnings to a library is a breaking change when the library is used with `PublishTrimm ...
+
+## Resolving trim warnings
```
How do you feel about combining @agocke's guidance blog here? https://github.com/mono/linker/pull/1946

There are good things about both descriptions. I'd rather have 1 really good one that combines the best of both, than have 2 different ones with different information.

sbomer replied:
I think it makes sense to have both a fixed announcement blog post and docs that we can expand over time. At the moment there's a lot of overlap, but I'd like to keep adding to these instructions so that they read more like a reference for authors who really have to dive into the annotations. @agocke thoughts?

I already see one discrepancy, which is that these docs don't yet mention UnconditionalSuppressMessage (I will add it). I'm also happy to take any other suggestions so that the docs contain a superset of the info in the blog post.

eerhardt replied:
I'm fine with 2 posts, as long as the official docs (i.e. this document) is better than the blog. The case I don't want to have happen is the blog to be really good, but the official docs being worse.

agocke replied:
I think this is good for long-term documentation. This looks more detailed, which I think is better for docs, and mine is more narrative, which may be useful for people just starting. I don't see any reason not to keep both up-to-date since the scenarios will be relevant long-term.

eerhardt commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    static void UseMethods(
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
```
I think it is better to start with a method that straight uses Reflection, teach how to annotate that method with `DynamicallyAccessedMembers`. Then the next step is to show other methods calling that first method, and how to annotate them.

eerhardt commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,188 @@
+
+### Recommendations
+
+In general, try to avoid reflection if possible, and when using reflection limit it in scope so that it is reachable only from a small part of the library. For example, avoid using non-understood patterns in places like static constructors that will result in the warning propagating to all members of the class.
```
```suggestion
In general, try to avoid reflection if possible. When using reflection, limit it in scope so that it is reachable only from a small part of the library. For example, avoid using non-understood patterns in places like static constructors that will result in the warning propagating to all members of the class.
```

eerhardt commented on docs/core/deploying/trimming-options.md:
```diff
@@ -9,27 +9,47 @@ ms.date: 08/25/2020
    Enable trimming during publish, with the default settings defined by the SDK.
 
-When using `Microsoft.NET.Sdk`, this will perform assembly-level trimming of the framework assemblies from the netcoreapp runtime pack. Application code and non-framework libraries are not trimmed. Other SDKs may define different defaults.
+When using `Microsoft.NET.Sdk`, this will trim the framework assemblies from the netcoreapp runtime pack. Application code and non-framework libraries are not trimmed. Other SDKs may define different defaults.
```
Not just `framework assemblies`, right? Any assembly marked with `IsTrimmable`, which all framework assemblies are annotated with.

sbomer replied:
Clarified this

eerhardt commented on docs/core/deploying/trimming-options.md:
```diff
@@ -9,27 +9,47 @@ ms.date: 08/25/2020
    Enable assembly-level trimming, which will keep an entire assembly if any part of it is used (in a statically understood way).
 
-- `<TrimMode>link</TrimMode>`
+Assemblies with `<IsTrimmable>true</IsTrimmable>` metadata but no explicit `TrimMode` will use the global `TrimMode`. The default `TrimMode` for `Microsoft.NET.Sdk` is `link` in .NET 6+, and `copyused` in previous versions.
```
> Assemblies with `<IsTrimmable>true</IsTrimmable>` metadata

What exactly does this mean? What assemblies? `TrimmableAssembly`? `ManagedAssemblyToLink`?

sbomer replied:
Clarified this in the section before the `TrimMode` bullets

eerhardt commented on docs/core/deploying/trimming-options.md:
```diff
@@ -114,6 +148,16 @@ Individual [warning codes](https://github.com/mono/linker/blob/master/docs/error
+
+- `<TrimmerSingleWarn>false</TrimmerSingleWarn>`
+
+    Show detailed warnings for all assemblies, instead of collapsing them to a single warning per assembly.
```
```suggestion
    Show all detailed warnings, instead of collapsing them to a single warning per assembly.
```

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

sbomer committed:
Update docs/core/deploying/trimming-options.md

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

sbomer committed:
PR feedback

- Fix links
- Clarify which assemblies are trimmed by default
- Clarify which assemblies are affected by per-assembly metadata

vitek-karas commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,188 @@
+
+These instructions show how to enable and resolve static analysis warnings to prepare a library for trimming. Follow these steps if you are authoring a library and either want to proactively make your library trimmable, or have been contacted by app authors who encountered trim warnings from your library.
+
+Use the .NET 6 SDK for the best experience. It is possible to [show analysis warnings](trimming-options.md#analysis-warnings) in .NET 5, but not all framework assemblies were annotated correctly in .NET 5. Also, this will show detailed warnings for every library, including framework libraries. These instructions assume you are using the .NET 6 SDK.
```
I would just state that users should use .NET 6 - and not go into the differences in behavior (there are way too many of them).

vitek-karas commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,188 @@
+
+### Incomplete Roslyn analyzer
+
+During development, you may set `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` (in .NET 6+) in your library project to get a _limited_ set of warnings from the Roslyn analyzer. This analyzer is incomplete and should only be used as a convenience. It is important to follow the next steps to ensure that your library is compatible with trimming.
```
This makes us look pretty bad - to soften this a bit I would add that we're working on improving the analyzer.

vitek-karas replied:
In fact I would probably change this section. (Rewrite in real english please :wink:)

### Enable analysis across toolchain

You should set `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` in your library project. This will not have any effect on the output, but it will enable trim analysis across the entire toolchain.

It will enable the Roslyn analyzer for trimming, which is currently not completely, but it's useful for the cases where it works. It provides immediate response from within IDE or build. It doesn't not report all trim analysis warnings and doesn't cover all patterns though. The analyzer should improve over time to provide wider coverage of the trim incompatible patterns.

sbomer replied:
Reworded, hopefully this is close to what you had in mind.

vitek-karas commented on docs/core/deploying/preparing-libraries-for-trimming.md:
````diff
@@ -0,0 +1,188 @@
+}
+```
+
+Here, `Foo` is calling a method which takes a `Type` argument that is annotated with a [`DynamicallyAccessedMembers`](https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute?view=net-5.0) requirement. The requirement states that the value passed in as this argument must represent a type whose public methods are available. In this case, you can fix ...
````
I would stay away from `Foo` - I know that at some point policheck wasn't happy about it. It also doesn't look "professional", or it looks "lazy". Basically just like we should not use `Foo` or similar simple name in tests - we should make the names descriptive of what they do.

vitek-karas commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,188 @@
+
+### Recommendations
+
+In general, try to avoid reflection if possible. When using reflection, limit it in scope so that it is reachable only from a small part of the library. For example, avoid using non-understood patterns in places like static constructors that will result in the warning propagating to all members of the class.
```
I would probably also mention virtual/interface methods as bad places to annotate (as it spreads very wide).

sbomer committed:
PR feedback

- Change example to call reflection directly
- Avoid mentioning .NET 5
- Point out benefits of Roslyn analyzer
- Avoid Foo/Bar
- Recommend not annotating virtuals

sbomer committed:
Add advanced section

Which mentions UnconditionalSuppressMessage

sbomer committed:
Remove whitespace

marek-safar commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,245 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
```
The description talks about library but output is exe

sbomer replied:
Clarified by improving the description before the sample

marek-safar commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,245 @@
+    <!-- Use a RID of your choice. -->
+    <RuntimeIdentifier>linux-x64</RuntimeIdentifier>
+    <PublishTrimmed>true</PublishTrimmed>
+    <TrimmerDefaultAction>link</TrimmerDefaultAction>
```
Is this necessary to have it here? Maybe just mention it bellow

sbomer replied:
I think it's pretty important - added a comment to the sample

marek-safar commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,245 @@
+    <RuntimeIdentifier>linux-x64</RuntimeIdentifier>
+    <PublishTrimmed>true</PublishTrimmed>
+    <TrimmerDefaultAction>link</TrimmerDefaultAction>
+    <!-- Optional, for non-default workloads -->
```
I don't understand what non-default workloads means

sbomer replied:
Fixed to describe what it does

marek-safar commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,245 @@
+
+  <ItemGroup>
+    <ProjectReference Include="path/to/MyLibrary.csproj" />
+    <TrimmerRootAssembly Include="MyLibrary" />
```
Isn't this the default behavior ?

sbomer replied:
It is, unless the library is trimmable. Added a comment inline to make it more clear

vitek-karas commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -22,11 +22,13 @@ If your app only uses parts of a library that are compatible with trimming, cons
+## Enable Roslyn analyzer
 
-During development, you may set `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` (in .NET 6+) in your library project to get a _limited_ set of warnings from the Roslyn analyzer. This analyzer is incomplete and should only be used as a convenience. It is important to follow the next steps to ensure that your library is compatible with trimming.
+During development, set `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` (in .NET 6+) in your library project. This will not have any effect on the output, but it will enable trim analysis during build via a Roslyn analyzer.
```
Why only during development?
If I want my library to be trim compatible - I should set this property... as simple as that.

sbomer committed:
PR feedback

- Clarify why we need an exe
- Improve comments in sample csproj

sbomer committed:
More comments

sbomer committed:
PR feedback

Simplify wording

IEvangelist commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,247 @@
+
+When publishing an app, `PublishTrimmed` will produce trim analysis [warnings](trimming-options.md#analysis-warnings) (in .NET 6+) for patterns that are not statically understood to be compatible with trimming, including patterns in your code and in dependencies.
+
+You will encounter detailed warnings originating from your own code and `ProjectReference` dependencies. You may also see warnings like `warning IL2104: Assembly 'SomeAssembly' produced trim warnings` for `PackageReference` libraries. This means that the library contained patterns which are not guaranteed to work in the context of the trimmed app, and may result in a broken app. Consider contacti ...
```
```suggestion
You will encounter detailed warnings originating from your own code and `ProjectReference` dependencies. You may also see warnings like `warning IL2104: Assembly 'SomeAssembly' produced trim warnings` for `PackageReference` libraries. This warning means that the library contained patterns which are not guaranteed to work in the context of the trimmed app, and may result in a broken app. Consider contacting the author to see if the library can be annotated for trimming.
```

IEvangelist commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,247 @@
+
+You will encounter detailed warnings originating from your own code and `ProjectReference` dependencies. You may also see warnings like `warning IL2104: Assembly 'SomeAssembly' produced trim warnings` for `PackageReference` libraries. This means that the library contained patterns which are not guaranteed to work in the context of the trimmed app, and may result in a broken app. Consider contacti ...
+
+To resolve warnings originating from the app code, jump ahead to the instructions on [resolving trim warnings](#resolving-trim-warnings). If you are interested in making your own `ProjectReference` libraries trim friendly, follow the instructions to [enable library trim warnings](#enabling-library-trim-warnings).
```
```suggestion
To resolve warnings originating from the app code, see [resolving trim warnings](#resolve-trim-warnings). If you are interested in making your own `ProjectReference` libraries trim friendly, follow the instructions to [enable library trim warnings](#enable-library-trim-warnings).
```

IEvangelist commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,247 @@
+
+You can also follow the same pattern for multiple libraries, adding them all to the same project as `ProjectReference` and `TrimmerRootAssembly` item to see trim analysis warnings for more than one library at a time, but note that this will warn about dependencies if _any_ of the root libraries use a trim-unfriendly API in a dependency. To see warnings that have to do with only a particular libra ...
+
+Note that the analysis results depend on the implementation details of your dependencies. If you update to a new version of a dependency, this may introduce analysis warnings if the new version added non-understood reflection patterns, even if there were no API changes. In other words, introducing trim analysis warnings to a library is a breaking change when the library is used with `PublishTrimm ...
```
```suggestion
> [!NOTE]
> The analysis results depend on the implementation details of your dependencies. If you update to a new version of a dependency, this may introduce analysis warnings if the new version added non-understood reflection patterns, even if there were no API changes. In other words, introducing trim analysis warnings to a library is a breaking change when the library is used with `PublishTrimmed`.
```

IEvangelist commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,247 @@
+
+Note that the analysis results depend on the implementation details of your dependencies. If you update to a new version of a dependency, this may introduce analysis warnings if the new version added non-understood reflection patterns, even if there were no API changes. In other words, introducing trim analysis warnings to a library is a breaking change when the library is used with `PublishTrimm ...
+
+## Resolving trim warnings
```
```suggestion
## Resolve trim warnings
```

IEvangelist reviewed (changes requested):
This is awesome, well done. I do have some suggestions I'd love to see incorporated into this to make it even better. Thanks

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/trimming-options.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/preparing-libraries-for-trimming.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Update docs/core/deploying/trimming-options.md

Co-authored-by: David Pine <david.pine@microsoft.com>

sbomer committed:
Add intro and fix links

sbomer commented on docs/core/deploying/preparing-libraries-for-trimming.md:
```diff
@@ -0,0 +1,247 @@
+---
+
+# Preparing libraries for trimming
+
```
Thanks, added a short intro!

sbomer committed:
Rename file to match title

marek-safar commented on docs/core/deploying/prepare-libraries-for-trimming.md:
```diff
@@ -0,0 +1,250 @@
+    <!-- Prevent warnings from unused code in dependencies -->
+    <TrimmerDefaultAction>link</TrimmerDefaultAction>
+    <!-- Enable warnings if the SDK you are using disables them -->
+    <SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>
```
I'm wondering if this should be done automatically, anytime when someone has `<TrimmerDefaultAction>link</TrimmerDefaultAction>`

sbomer replied:
Removed, since this will be done automatically in https://github.com/dotnet/sdk/pull/16865.

marek-safar commented on docs/core/deploying/prepare-libraries-for-trimming.md:
```diff
@@ -0,0 +1,250 @@
+
+The Roslyn analyzer is useful for a fast feedback cycle with IDE integration, but is currently incomplete. It doesn't cover all trim analysis warnings, but the set of patterns it understands will improve over time to give more complete coverage. The Roslyn analyzer also isn't able to analyze the implementations of reference assemblies that you depend on. It is important to follow the next steps t ...
+
+### Show all warnings
```
This section is not really only about showing all warnings but about trimming everything, right? Should this section cover how to do it for a library and not only for an app?

sbomer replied:
It is about showing library warnings - I added a sentence to clarify why the step of creating an app project is necessary.

IEvangelist commented on docs/core/deploying/prepare-libraries-for-trimming.md:
````diff
@@ -0,0 +1,250 @@
+```
+
+This means the library calls a method which has explicitly been annotated as incompatible with trimming, using [`RequiresUnreferencedCodeAttribute`](
+https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.requiresunreferencedcodeattribute?view=net-5.0). To get rid of the warning, consider whether `Method` needs to call `DynamicBehavior` to do its job. If so, annotate the caller `Method` with `RequiresUnreferencedCode` as well; this will "bubble up" the warning so that callers of `Method` get a warning instead:
````
```suggestion
https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.requiresunreferencedcodeattribute?view=net-5.0&preserve-view=true). To get rid of the warning, consider whether `Method` needs to call `DynamicBehavior` to do its job. If so, annotate the caller `Method` with `RequiresUnreferencedCode` as well; this will "bubble up" the warning so that callers of `Method` get a warning instead:
```

IEvangelist commented on docs/core/deploying/prepare-libraries-for-trimming.md:
````diff
@@ -0,0 +1,250 @@
+}
+```
+
+Here, `UseMethods` is calling a reflection method which has a [`DynamicallyAccessedMembers`](https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute?view=net-5.0) requirement. The requirement states that the type's public methods are available. In this case, you can fix this by adding the same requirement to the parameter of `UseMethods`.
````
```suggestion
Here, `UseMethods` is calling a reflection method which has a [`DynamicallyAccessedMembers`](https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute?view=net-5.0&preserve-view=true) requirement. The requirement states that the type's public methods are available. In this case, you can fix this by adding the same requirement to the parameter of `UseMethods`.
```

IEvangelist commented on docs/core/deploying/prepare-libraries-for-trimming.md:
```diff
@@ -0,0 +1,250 @@
+
+### UnconditionalSuppressMessage
+
+If the intent of your code can't be expressed with the annotations, but you know that the warning doesn't represent a real issue at runtime, you can suppress the warnings using [`UnconditionalSuppressMessageAttribute`](https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.unconditionalsuppressmessageattribute?view=net-5.0). This is similar to `SuppressMessageAttribute`, but it is  ...
```
```suggestion
If the intent of your code can't be expressed with the annotations, but you know that the warning doesn't represent a real issue at runtime, you can suppress the warnings using [`UnconditionalSuppressMessageAttribute`](https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.unconditionalsuppressmessageattribute?view=net-5.0&preserve-view=true). This is similar to `SuppressMessageAttribute`, but it is persisted in IL and respected during trim analysis. For example:
```

IEvangelist committed:
Update trimming-options.md

IEvangelist committed:
Apply suggestions from code review

IEvangelist reviewed (approved):
LGTM, we'll :shipit: if everyone else is good with this - thank you 🙏

sbomer committed:
PR feedback

- Don't require un-suppressing warnings, as this is now implied by TrimmerDefaultAction=link:
https://github.com/dotnet/sdk/pull/16865
- Clarify why publishing an app is necessary for library warnings

gewarren commented:
Close/reopen to test status check.
